File: python/src/lichen/coap/resources/rangetest.py
# SPDX-License-Identifier: GPL-3.0-or-later
# SPDX-FileCopyrightText: The contributors to the LICHEN project
"""Range Testing resources: /diag/rangetest and /diag/traceroute (spec 18.7)."""

from __future__ import annotations

import math
import time
from dataclasses import dataclass, field
from typing import Any

import aiocoap
import cbor2
from aiocoap import CONTENT, Message, resource

from lichen.coap.resources.base import CBOR, SENML_CBOR
from lichen.coap.resources.cbor_validation import _decode_single_cbor
from lichen.senml.codec import SenmlRecord, pack

# Limits per spec
MAX_PAYLOAD_LEN = 255  # Maximum test payload size
MAX_COUNT = 100  # Maximum response count
DEFAULT_INTERVAL_MS = 5000  # Default continuous test interval


@dataclass
class RadioMetrics:
    """Radio link quality metrics for range testing."""

    rssi: float = -85.0  # dBm
    snr: float = 7.5  # dB
    sf: int = 9  # Spreading factor
    freq: float = 906.875  # MHz


@dataclass
class TracerouteHop:
    """A single hop in a traceroute."""

    addr: str  # IPv6 address
    rssi: float  # dBm
    rtt_ms: float  # Round-trip time in ms


@dataclass
class RadioMetricsProvider:
    """Provider for radio metrics (injectable for testing).

    In a real implementation, this would read from the radio driver.
    """

    metrics: RadioMetrics = field(default_factory=RadioMetrics)
    hops: list[TracerouteHop] = field(default_factory=list)
    node_eui64: str = "0102030405060708"

    def get_metrics(self) -> RadioMetrics:
        """Return current radio metrics."""
        return self.metrics

    def get_traceroute(self) -> list[TracerouteHop]:
        """Return traceroute hops to this node."""
        return self.hops


def _rangetest_senml(
    node_eui64: str,
    seq: int,
    metrics: RadioMetrics,
    timestamp: float | None = None,
) -> list[SenmlRecord]:
    """Build SenML pack for range test response per spec 18.7.2.

    Returns:
        List of SenML records with bn, bt, seq, rssi, snr, sf, freq.
    """
    bt = timestamp if timestamp is not None else time.time()
    return [
        SenmlRecord(bn=f"urn:dev:mac:{node_eui64}:", bt=bt),
        SenmlRecord(n="seq", v=seq),
        SenmlRecord(n="rssi", u="dBm", v=metrics.rssi),
        SenmlRecord(n="snr", u="dB", v=metrics.snr),
        SenmlRecord(n="sf", v=metrics.sf),
        SenmlRecord(n="freq", u="MHz", v=metrics.freq),
    ]


class RangeTestResource(resource.ObservableResource):
    """Observable ``/diag/rangetest`` — range testing (spec 18.7.2/18.7.3).

    **POST** (Extended Range Test): Returns SenML+CBOR with radio metrics.
    Request body (optional)::

        {"seq": 1, "payload_len": 32, "count": 5}

    Response::

        [
          {"bn": "urn:dev:mac:...", "bt": 1716742800},
          {"n": "seq", "v": 1},
          {"n": "rssi", "u": "dBm", "v": -85},
          {"n": "snr", "u": "dB", "v": 7.5},
          {"n": "sf", "v": 9},
          {"n": "freq", "u": "MHz", "v": 906.875}
        ]

    **GET** (Continuous Range Test): Returns current metrics or subscribes
    via Observe (RFC 7641) for periodic updates.

    The resource maintains internal state that can be updated via :meth:`update`
    to push new readings to observers.
    """

    def __init__(
        self,
        provider: RadioMetricsProvider | None = None,
        time_func: Any = None,
    ) -> None:
        """Initialize RangeTest resource.

        Args:
            provider: Radio metrics provider (defaults to static metrics).
            time_func: Optional callable returning current time (for testing).
        """
        super().__init__()
        self._provider = provider if provider is not None else RadioMetricsProvider()
        self._time_func = time_func if time_func is not None else time.time
        self._seq = 0
        self._interval_ms = DEFAULT_INTERVAL_MS
        # Pre-compute initial payload
        self._update_payload()

    def _update_payload(self) -> None:
        """Update the cached SenML payload."""
        metrics = self._provider.get_metrics()
        records = _rangetest_senml(
            self._provider.node_eui64,
            self._seq,
            metrics,
            self._time_func(),
        )
        self._payload = pack(records)

    def update(self, metrics: RadioMetrics | None = None) -> None:
        """Update metrics and notify observers.

        Args:
            metrics: New radio metrics. If None, re-reads from provider.
        """
        if metrics is not None:
            self._provider.metrics = metrics
        self._seq += 1
        self._update_payload()
        self.updated_state()

    def get_link_description(self) -> dict[str, Any]:
        """Link description for .well-known/core."""
        return {
            "rt": "rangetest",
            "ct": str(int(SENML_CBOR)),
            "obs": None,
        }

    async def render_get(self, request: Message) -> Message:
        """GET /diag/rangetest — return current metrics (spec 18.7.3)."""
        # Check for interval parameter in payload
        if request.payload:
            try:
                body = _decode_single_cbor(request.payload)
            except (ValueError, OverflowError, cbor2.CBORDecodeError):
                return Message(code=aiocoap.BAD_REQUEST)
            # Match POST strictness: only a map may carry options (spec 18.7.3).
            if not isinstance(body, dict):
                return Message(code=aiocoap.BAD_REQUEST)
            if "interval_ms" in body:
                interval = body["interval_ms"]
                if (
                    isinstance(interval, bool)
                    or not isinstance(interval, (int, float))
                    or (isinstance(interval, float) and not math.isfinite(interval))
                    or interval <= 0
                ):
                    return Message(code=aiocoap.BAD_REQUEST)
                self._interval_ms = int(interval)

        # Update and return current metrics
        self._update_payload()
        msg = Message(code=CONTENT, payload=self._payload)
        msg.opt.content_format = SENML_CBOR
        return msg

    async def render_post(self, request: Message) -> Message:
        """POST /diag/rangetest — extended range test (spec 18.7.2)."""
        seq = 0
        # payload_len and count are validated but behavior is out of scope
        # for this oracle; real implementation would pad request/send multiple responses
        _payload_len = 0
        _count = 1

        # Parse optional request body
        if request.payload:
            try:
                body = _decode_single_cbor(request.payload)
            except (ValueError, OverflowError, cbor2.CBORDecodeError):
                return Message(code=aiocoap.BAD_REQUEST)

            if not isinstance(body, dict):
                return Message(code=aiocoap.BAD_REQUEST)

            # Validate seq
            if "seq" in body:
                seq_val = body["seq"]
                if isinstance(seq_val, bool) or not isinstance(seq_val, int) or seq_val < 0:
                    return Message(code=aiocoap.BAD_REQUEST)
                seq = seq_val

            # Validate payload_len
            if "payload_len" in body:
                plen = body["payload_len"]
                if (
                    isinstance(plen, bool)
                    or not isinstance(plen, int)
                    or plen < 0
                    or plen > MAX_PAYLOAD_LEN
                ):
                    return Message(code=aiocoap.BAD_REQUEST)
                _payload_len = plen  # noqa: F841

            # Validate count
            if "count" in body:
                cnt = body["count"]
                if isinstance(cnt, bool) or not isinstance(cnt, int) or cnt < 1 or cnt > MAX_COUNT:
                    return Message(code=aiocoap.BAD_REQUEST)
                _count = cnt  # noqa: F841

        # Get current metrics from provider
        metrics = self._provider.get_metrics()

        # Build SenML response
        records = _rangetest_senml(
            self._provider.node_eui64,
            seq,
            metrics,
            self._time_func(),
        )

        # For count > 1, the caller would receive multiple responses
        # (out of scope for this oracle; just return single response)
        # The payload_len parameter is for testing with specific sizes
        # (would affect the request padding, not the response)

        payload = pack(records)
        msg = Message(code=CONTENT, payload=payload)
        msg.opt.content_format = SENML_CBOR
        return msg


class TracerouteResource(resource.Resource):
    """``/diag/traceroute`` — mesh path discovery (spec 18.7.4).

    **GET** returns the hop-by-hop path through the mesh::

        {
          "hops": [
            {"addr": "fe80::1111", "rssi": -65, "rtt_ms": 120},
            {"addr": "fe80::2222", "rssi": -78, "rtt_ms": 340}
          ],
          "total_hops": 2,
          "total_rtt_ms": 340
        }

    Implementation uses RPL source routing information or hop-by-hop probing.
    """

    def __init__(self, provider: RadioMetricsProvider | None = None) -> None:
        """Initialize Traceroute resource.

        Args:
            provider: Radio metrics provider with hop information.
        """
        super().__init__()
        self._provider = provider if provider is not None else RadioMetricsProvider()

    def get_link_description(self) -> dict[str, Any]:
        """Link description for .well-known/core."""
        return {
            "rt": "traceroute",
            "ct": str(int(CBOR)),
        }

    async def render_get(self, request: Message) -> Message:
        """GET /diag/traceroute — return mesh path (spec 18.7.4)."""
        hops = self._provider.get_traceroute()

        # Build response
        hops_data = [{"addr": hop.addr, "rssi": hop.rssi, "rtt_ms": hop.rtt_ms} for hop in hops]

        total_hops = len(hops)
        # Float-typed even when empty so the CBOR value type never flips
        # with hop count (cross-implementation conformance).
        total_rtt_ms = hops[-1].rtt_ms if hops else 0.0

        response = {
            "hops": hops_data,
            "total_hops": total_hops,
            "total_rtt_ms": total_rtt_ms,
        }

        msg = Message(code=CONTENT, payload=cbor2.dumps(response))
        msg.opt.content_format = CBOR
        return msg

File: rust/lichen-client/src/rangetest.rs
// SPDX-License-Identifier: GPL-3.0-or-later
// SPDX-FileCopyrightText: The contributors to the LICHEN project

//! Range testing domain types and CBOR wire codecs (spec 18.7).
//!
//! Wire contract (firmware `/diag/rangetest` and `/diag/traceroute`
//! resources):
//!
//! * Extended/continuous range test responses are SenML+CBOR packs
//!   (RFC 8428, Content-Format 112) with numeric labels: `bn` = -2,
//!   `bt` = -3, `n` = 0, `u` = 1, `v` = 2. Records carry `seq`, `rssi`
//!   (dBm), `snr` (dB), `sf`, and `freq` (MHz).
//! * Traceroute responses are a CBOR map with text keys
//!   `hops` / `total_hops` / `total_rtt_ms`.
//! * Request bodies are CBOR maps with text keys: `seq`, `payload_len`,
//!   `count` for POST; `interval_ms` for GET.
//!
//! Conformance vectors: `test/vectors/rangetest.json`.

use ciborium::value::Value;

use crate::Error;

/// Maximum test payload size per spec 18.7.2.
pub const MAX_PAYLOAD_LEN: u32 = 255;
/// Maximum response count per spec 18.7.2.
pub const MAX_COUNT: u32 = 100;
/// Default continuous-test interval in milliseconds.
pub const DEFAULT_INTERVAL_MS: u64 = 5000;

/// Radio link quality metrics reported by a range test.
#[derive(Debug, Clone, PartialEq)]
pub struct RadioMetrics {
    /// Received signal strength in dBm (negative).
    pub rssi: f64,
    /// Signal-to-noise ratio in dB.
    pub snr: f64,
    /// LoRa spreading factor.
    pub sf: u8,
    /// Center frequency in MHz.
    pub freq: f64,
}

/// One decoded range-test reading (SenML pack).
#[derive(Debug, Clone, PartialEq)]
pub struct RangeTestReading {
    /// Base name (`bn`), e.g. `urn:dev:mac:<eui64>:`.
    pub base_name: Option<String>,
    /// Base time (`bt`) in Unix seconds.
    pub base_time: Option<f64>,
    /// Test sequence number echoed by the responder.
    pub seq: i64,
    /// Current radio metrics.
    pub metrics: RadioMetrics,
}

/// A single hop in a mesh traceroute.
#[derive(Debug, Clone, PartialEq)]
pub struct TracerouteHop {
    /// Next-hop IPv6 address.
    pub addr: String,
    /// Link RSSI toward that hop in dBm.
    pub rssi: f64,
    /// Round-trip time to that hop in milliseconds.
    pub rtt_ms: f64,
}

/// Decoded `/diag/traceroute` response.
#[derive(Debug, Clone, PartialEq)]
pub struct TracerouteResult {
    /// Hop-by-hop path.
    pub hops: Vec<TracerouteHop>,
    /// Number of hops (equal to `hops.len()`).
    pub total_hops: usize,
    /// End-to-end round trip in milliseconds (last hop's RTT).
    pub total_rtt_ms: f64,
}

/// Extended range-test request body (`POST /diag/rangetest`).
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct RangeTestRequest {
    /// Sequence number to echo (>= 0).
    pub seq: Option<u32>,
    /// Padding payload size in bytes (0..=[MAX_PAYLOAD_LEN]).
    pub payload_len: Option<u32>,
    /// Number of exchanges (1..=[MAX_COUNT]).
    pub count: Option<u32>,
}

fn value_label(map: &[(Value, Value)], key: i64) -> Option<&Value> {
    map.iter()
        .find(|(k, _)| matches!(k, Value::Integer(i) if i64::try_from(*i).is_ok_and(|v| v == key)))
        .map(|(_, v)| v)
}

fn text_key<'a>(map: &'a [(Value, Value)], key: &str) -> Option<&'a Value> {
    map.iter()
        .find(|(k, _)| matches!(k, Value::Text(t) if t == key))
        .map(|(_, v)| v)
}

fn as_i64(v: &Value) -> Result<i64, Error> {
    match v {
        Value::Integer(i) => {
            i64::try_from(*i).map_err(|_| Error::Decode("integer out of range".into()))
        }
        _ => Err(Error::Decode(format!("expected integer, got {v:?}"))),
    }
}

fn as_f64(v: &Value) -> Result<f64, Error> {
    match v {
        Value::Integer(i) => match i64::try_from(*i) {
            Ok(v) => Ok(v as f64),
            Err(_) => Err(Error::Decode("integer too large for float".into())),
        },
        Value::Float(f) => Ok(*f),
        _ => Err(Error::Decode(format!("expected number, got {v:?}"))),
    }
}

fn as_map(v: &Value) -> Result<&Vec<(Value, Value)>, Error> {
    match v {
        Value::Map(m) => Ok(m),
        _ => Err(Error::Decode(format!("expected map, got {v:?}"))),
    }
}

fn as_array(v: &Value) -> Result<&Vec<Value>, Error> {
    match v {
        Value::Array(a) => Ok(a),
        _ => Err(Error::Decode(format!("expected array, got {v:?}"))),
    }
}

impl RangeTestReading {
    /// Decode a SenML+CBOR range-test response pack.
    pub fn from_senml_cbor(bytes: &[u8]) -> Result<Self, Error> {
        let value: Value =
            ciborium::from_reader(bytes).map_err(|e| Error::Decode(e.to_string()))?;
        let records = as_array(&value)?;

        let mut base_name = None;
        let mut base_time = None;
        let mut seq = None;
        let mut rssi = None;
        let mut snr = None;
        let mut sf = None;
        let mut freq = None;

        for record in records {
            let map = as_map(record)?;
            let name = value_label(map, 0)
                .and_then(Value::as_text)
                .map(str::to_owned);
            match name.as_deref() {
                Some("seq") => {
                    seq = Some(as_i64(value_label(map, 2).ok_or_else(|| {
                        Error::Decode("seq record missing value".into())
                    })?)?)
                }
                Some("rssi") | Some("snr") | Some("freq") => {
                    let v = as_f64(
                        value_label(map, 2)
                            .ok_or_else(|| Error::Decode("metric record missing value".into()))?,
                    )?;
                    match name.as_deref() {
                        Some("rssi") => rssi = Some(v),
                        Some("snr") => snr = Some(v),
                        _ => freq = Some(v),
                    }
                }
                Some("sf") => {
                    sf = Some(
                        u8::try_from(as_i64(
                            value_label(map, 2)
                                .ok_or_else(|| Error::Decode("sf record missing value".into()))?,
                        )?)
                        .map_err(|_| Error::Decode("sf out of u8 range".into()))?,
                    );
                }
                _ => {
                    // Base record and unknown records: harvest base fields only.
                    if let Some(v) = value_label(map, -2) {
                        base_name = Some(
                            v.as_text()
                                .ok_or_else(|| Error::Decode("bn must be text".into()))?
                                .to_owned(),
                        );
                    }
                    if let Some(v) = value_label(map, -3) {
                        base_time = Some(as_f64(v)?);
                    }
                }
            }
        }

        Ok(Self {
            base_name,
            base_time,
            seq: seq.ok_or_else(|| Error::Decode("missing seq record".into()))?,
            metrics: RadioMetrics {
                rssi: rssi.ok_or_else(|| Error::Decode("missing rssi record".into()))?,
                snr: snr.ok_or_else(|| Error::Decode("missing snr record".into()))?,
                sf: sf.ok_or_else(|| Error::Decode("missing sf record".into()))?,
                freq: freq.ok_or_else(|| Error::Decode("missing freq record".into()))?,
            },
        })
    }
}

impl RangeTestRequest {
    fn validate(&self) -> Result<(), Error> {
        if let Some(plen) = self.payload_len {
            if plen > MAX_PAYLOAD_LEN {
                return Err(Error::Encode(format!(
                    "payload_len {plen} exceeds MAX_PAYLOAD_LEN ({MAX_PAYLOAD_LEN})"
                )));
            }
        }
        if let Some(count) = self.count {
            if !(1..=MAX_COUNT).contains(&count) {
                return Err(Error::Encode(format!(
                    "count {count} outside 1..={MAX_COUNT}"
                )));
            }
        }
        Ok(())
    }

    /// Encode the request body as a CBOR map with text keys.
    ///
    /// Only set fields are emitted; bounds per spec 18.7.2 are enforced.
    pub fn to_cbor(&self) -> Result<Vec<u8>, Error> {
        self.validate()?;
        let mut entries = Vec::new();
        if let Some(seq) = self.seq {
            entries.push((Value::Text("seq".into()), Value::Integer(seq.into())));
        }
        if let Some(plen) = self.payload_len {
            entries.push((
                Value::Text("payload_len".into()),
                Value::Integer(plen.into()),
            ));
        }
        if let Some(count) = self.count {
            entries.push((Value::Text("count".into()), Value::Integer(count.into())));
        }
        let mut bytes = Vec::new();
        ciborium::into_writer(&Value::Map(entries), &mut bytes)
            .map_err(|e| Error::Encode(e.to_string()))?;
        Ok(bytes)
    }

    /// Decode a request body (used when echoing or inspecting requests).
    pub fn from_cbor(bytes: &[u8]) -> Result<Self, Error> {
        let value: Value =
            ciborium::from_reader(bytes).map_err(|e| Error::Decode(e.to_string()))?;
        let map = as_map(&value)?;
        let get_u32 = |key: &str| -> Result<Option<u32>, Error> {
            match text_key(map, key) {
                Some(v) => {
                    Ok(Some(u32::try_from(as_i64(v)?).map_err(|_| {
                        Error::Decode(format!("{key} out of u32 range"))
                    })?))
                }
                None => Ok(None),
            }
        };
        Ok(Self {
            seq: get_u32("seq")?,
            payload_len: get_u32("payload_len")?,
            count: get_u32("count")?,
        })
    }
}

/// Encode a continuous-test interval body (`GET /diag/rangetest`).
///
/// The interval must be positive; the server rejects anything else with 4.00.
pub fn interval_body(interval_ms: u64) -> Result<Vec<u8>, Error> {
    if interval_ms == 0 {
        return Err(Error::Encode("interval_ms must be positive".into()));
    }
    let mut bytes = Vec::new();
    ciborium::into_writer(
        &Value::Map(vec![(
            Value::Text("interval_ms".into()),
            Value::Integer(interval_ms.into()),
        )]),
        &mut bytes,
    )
    .map_err(|e| Error::Encode(e.to_string()))?;
    Ok(bytes)
}

/// Minimal deterministic CBOR writers matching the reference codec's output
/// (Python `cbor2.dumps` emits 64-bit floats, never half/single precision).
mod wire {
    pub(super) fn uint(out: &mut Vec<u8>, v: u64) {
        if v < 24 {
            out.push(v as u8);
        } else if v <= u64::from(u8::MAX) {
            out.push(24);
            out.push(v as u8);
        } else if v <= u64::from(u16::MAX) {
            out.push(25);
            out.extend_from_slice(&(v as u16).to_be_bytes());
        } else if v <= u64::from(u32::MAX) {
            out.push(26);
            out.extend_from_slice(&(v as u32).to_be_bytes());
        } else {
            out.push(27);
            out.extend_from_slice(&v.to_be_bytes());
        }
    }

    pub(super) fn text(out: &mut Vec<u8>, s: &str) {
        let len = s.len();
        if len < 24 {
            out.push(0x60 | len as u8);
        } else if len <= usize::from(u8::MAX) {
            out.push(0x78);
            out.push(len as u8);
        } else if len <= usize::from(u16::MAX) {
            out.push(0x79);
            out.extend_from_slice(&(len as u16).to_be_bytes());
        } else {
            out.push(0x7A);
            out.extend_from_slice(&(len as u32).to_be_bytes());
        }
        out.extend_from_slice(s.as_bytes());
    }

    pub(super) fn array(out: &mut Vec<u8>, len: usize) {
        debug_assert!(len <= u32::MAX as usize);
        if len < 24 {
            out.push(0x80 | len as u8);
        } else if len <= usize::from(u8::MAX) {
            out.push(0x98);
            out.push(len as u8);
        } else {
            out.push(0x99);
            out.extend_from_slice(&(len as u16).to_be_bytes());
        }
    }

    pub(super) fn map(out: &mut Vec<u8>, len: usize) {
        debug_assert!(len <= 23, "traceroute maps are fixed-shape");
        out.push(0xA0 | len as u8);
    }

    pub(super) fn float64(out: &mut Vec<u8>, v: f64) {
        out.push(0xFB);
        out.extend_from_slice(&v.to_be_bytes());
    }
}

impl TracerouteResult {
    /// Decode a `/diag/traceroute` CBOR response.
    pub fn from_cbor(bytes: &[u8]) -> Result<Self, Error> {
        let value: Value =
            ciborium::from_reader(bytes).map_err(|e| Error::Decode(e.to_string()))?;
        let map = as_map(&value)?;

        let hops_value =
            as_array(text_key(map, "hops").ok_or_else(|| Error::Decode("missing hops".into()))?)?;
        let mut hops = Vec::with_capacity(hops_value.len());
        for hop in hops_value {
            let hop_map = as_map(hop)?;
            let addr = text_key(hop_map, "addr")
                .ok_or_else(|| Error::Decode("hop missing addr".into()))?
                .as_text()
                .ok_or_else(|| Error::Decode("addr must be text".into()))?
                .to_owned();
            let rssi = as_f64(
                text_key(hop_map, "rssi")
                    .ok_or_else(|| Error::Decode("hop missing rssi".into()))?,
            )?;
            let rtt_ms = as_f64(
                text_key(hop_map, "rtt_ms")
                    .ok_or_else(|| Error::Decode("hop missing rtt_ms".into()))?,
            )?;
            hops.push(TracerouteHop { addr, rssi, rtt_ms });
        }

        let total_hops = usize::try_from(as_i64(
            text_key(map, "total_hops")
                .ok_or_else(|| Error::Decode("missing total_hops".into()))?,
        )?)
        .map_err(|_| Error::Decode("total_hops out of range".into()))?;

        let total_rtt_ms = as_f64(
            text_key(map, "total_rtt_ms")
                .ok_or_else(|| Error::Decode("missing total_rtt_ms".into()))?,
        )?;

        Ok(Self {
            hops,
            total_hops,
            total_rtt_ms,
        })
    }

    /// Encode a traceroute response body (text-keyed map, field order
    /// `hops`, `total_hops`, `total_rtt_ms` matching the reference codec).
    ///
    /// Floats are emitted as 64-bit to match the Python reference bytes
    /// (ciborium's minimal-width float output would break byte-exact
    /// conformance).
    pub fn to_cbor(&self) -> Result<Vec<u8>, Error> {
        let mut out = Vec::new();
        wire::map(&mut out, 3);
        wire::text(&mut out, "hops");
        wire::array(&mut out, self.hops.len());
        for h in &self.hops {
            wire::map(&mut out, 3);
            wire::text(&mut out, "addr");
            wire::text(&mut out, &h.addr);
            wire::text(&mut out, "rssi");
            wire::float64(&mut out, h.rssi);
            wire::text(&mut out, "rtt_ms");
            wire::float64(&mut out, h.rtt_ms);
        }
        wire::text(&mut out, "total_hops");
        wire::uint(
            &mut out,
            u64::try_from(self.total_hops)
                .map_err(|_| Error::Encode("total_hops out of range".into()))?,
        );
        wire::text(&mut out, "total_rtt_ms");
        wire::float64(&mut out, self.total_rtt_ms);
        Ok(out)
    }
}
